zeppelin-server/src/main/java/org/apache/zeppelin/rest/NotebookRestApi.java

```diff
@@ -567,22 +568,67 @@ public Response runParagraph(@PathParam("notebookId") String notebookId,
       return new JsonResponse<>(Status.NOT_FOUND, "paragraph not found.").build();
     }
 
-    // handle params if presented
-    if (!StringUtils.isEmpty(message)) {
-      RunParagraphWithParametersRequest request = gson.fromJson(message,
-          RunParagraphWithParametersRequest.class);
-      Map<String, Object> paramsForUpdating = request.getParams();
-      if (paramsForUpdating != null) {
-        paragraph.settings.getParams().putAll(paramsForUpdating);
-        note.persist();
-      }
-    }
+    // handle params if presented 
+    handleParagraphParams(message, note, paragraph);
+
 
     note.run(paragraph.getId());
     return new JsonResponse<>(Status.OK).build();
   }
 
   /**
+   * Run synchronously a paragraph REST API
+   *
+   * @param noteId - noteId
+   * @param paragraphId - paragraphId
+   * @param message - JSON with params if user wants to update dynamic form's value
+   *                null, empty string, empty json if user doesn't want to update
+   *
+   * @return JSON with status.OK
+   * @throws IOException, IllegalArgumentException
+   */
+  @POST
+  @Path("run/{notebookId}/{paragraphId}")
+  @ZeppelinApi
+  public Response runParagraphSynchronously(@PathParam("notebookId") String noteId,
+                                            @PathParam("paragraphId") String paragraphId,
+                                            String message) throws
+                                            IOException, IllegalArgumentException {
+    LOG.info("run paragraph synchronously {} {} {}", noteId, paragraphId, message);
+
+    Note note = notebook.getNote(noteId);
+    if (note == null) {
+      return new JsonResponse<>(Status.NOT_FOUND, "note not found.").build();
+    }
+
+    Paragraph paragraph = note.getParagraph(paragraphId);
+    if (paragraph == null) {
+      return new JsonResponse<>(Status.NOT_FOUND, "paragraph not found.").build();
+    }
+
+    // handle params if presented
+    handleParagraphParams(message, note, paragraph);
+
+    if (paragraph.getNoteReplLoader() == null) {
+      paragraph.setNoteReplLoader(note.getNoteReplLoader());
+    }
+
+    if (paragraph.getListener() == null) {
+      paragraph.setListener(note.getJobListenerFactory().getParagraphJobListener(note));
+    }
+
+    paragraph.run();
+
+    final InterpreterResult result = paragraph.getResult();
+
+    if (result.code() == InterpreterResult.Code.SUCCESS) {
+      return new JsonResponse<>(Status.OK, result).build();
+    } else {
+      return new JsonResponse<>(Status.INTERNAL_SERVER_ERROR, result).build();
```

Comment:
Well, the `InterpreterResult.Code` enum has only 4 distinct values right now:

1. SUCCESS,
2. INCOMPLETE
3. ERROR
4. KEEP_PREVIOUS_RESULT

Success case is handle in the  `if` block. ERROR can be mapped to `Status.INTERNAL_SERVER_ERROR`. INCOMPLETE and KEEP_PREVIOUS_RESULT is Zeppelin own internal state to manage streaming from interpreter so we cannot map them to a sensible HTTP error code.

In any case, since we return the complete `InterpreterResult` object, we also pass the error detail back to the client

Example:

![image](https://cloud.githubusercontent.com/assets/1532977/15953514/4559a92e-2e91-11e6-91da-2c8be1dba077.png)

```
{
    "status": "INTERNAL_SERVER_ERROR",
    "body": {
        "code": "ERROR",
        "type": "TEXT",
        "msg": "cat: x: No such file or directory\nExitValue: 1"
    }
}
```